x/emissions/abci.go

```diff
@@ -35,71 +37,92 @@ func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper, stakingKeeper types.Sta
-	coin := sdk.NewCoins(sdk.NewCoin(config.BaseDenom, amount))
-	return bankKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, types.UndistributedObserverRewardsPool, coin)
-}
-
-func GetBlockRewardComponents(ctx sdk.Context, bankKeeper types.BankKeeper, stakingKeeper types.StakingKeeper, emissionKeeper keeper.Keeper) (sdk.Dec, sdk.Dec, sdk.Dec) {
-	reservesFactor := GetReservesFactor(ctx, bankKeeper)
-	if reservesFactor.LTE(sdk.ZeroDec()) {
-		return sdk.ZeroDec(), sdk.ZeroDec(), sdk.ZeroDec()
-	}
-	bondFactor := GetBondFactor(ctx, stakingKeeper, emissionKeeper)
-	durationFactor := GetDurationFactor(ctx, emissionKeeper)
-	return reservesFactor, bondFactor, durationFactor
-}
+// DistributeObserverRewards distributes the rewards to all observers who voted in any of the matured ballots
+// The total rewards are distributed equally among all Successful votes
+// NotVoted or Unsuccessful votes are slashed
+// rewards given or slashed amounts are in azeta
 
-func DistributeTssRewards(ctx sdk.Context, amount sdk.Int, bankKeeper types.BankKeeper) error {
-	coin := sdk.NewCoins(sdk.NewCoin(config.BaseDenom, amount))
-	return bankKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, types.UndistributedTssRewardsPool, coin)
-}
-func GetBondFactor(ctx sdk.Context, stakingKeeper types.StakingKeeper, keeper keeper.Keeper) sdk.Dec {
-	targetBondRatio := sdk.MustNewDecFromStr(keeper.GetParams(ctx).TargetBondRatio)
-	maxBondFactor := sdk.MustNewDecFromStr(keeper.GetParams(ctx).MaxBondFactor)
-	minBondFactor := sdk.MustNewDecFromStr(keeper.GetParams(ctx).MinBondFactor)
+func DistributeObserverRewards(ctx sdk.Context, amount sdkmath.Int, keeper keeper.Keeper) error {
 
-	currentBondedRatio := stakingKeeper.BondedRatio(ctx)
-	// Bond factor ranges between minBondFactor (0.75) to maxBondFactor (1.25)
-	if currentBondedRatio.IsZero() {
-		return sdk.ZeroDec()
+	rewardsDistributer := map[string]int64{}
+	totalRewardsUnits := int64(0)
+	err := keeper.GetBankKeeper().SendCoinsFromModuleToModule(ctx, types.ModuleName, types.UndistributedObserverRewardsPool, sdk.NewCoins(sdk.NewCoin(config.BaseDenom, amount)))
+	if err != nil {
+		return err
 	}
-	bondFactor := targetBondRatio.Quo(currentBondedRatio)
-	if bondFactor.GT(maxBondFactor) {
-		return maxBondFactor
+	ballotIdentifiers := keeper.GetObserverKeeper().GetMaturedBallotList(ctx)
+	// do not distribute rewards if no ballots are matured, the rewards can accumulate in the undistributed pool
+	if len(ballotIdentifiers) == 0 {
+		return nil
 	}
-	if bondFactor.LT(minBondFactor) {
-		return minBondFactor
+	for _, ballotIdentifier := range ballotIdentifiers {
+		ballot, found := keeper.GetObserverKeeper().GetBallot(ctx, ballotIdentifier)
+		if !found {
+			continue
+		}
+		totalRewardsUnits = totalRewardsUnits + ballot.BuildRewardsDistribution(rewardsDistributer)
+	}
+	rewardPerUnit := sdkmath.ZeroInt()
+	if totalRewardsUnits > 0 && amount.IsPositive() {
+		rewardPerUnit = amount.Quo(sdk.NewInt(totalRewardsUnits))
```

Comment:
Ballots can be cleared after this distribution if needed, decided not to do this in this PR . Can be added in later